Exiled.Events/Patches/Events/Player/ExitingSinkholeEnvironmentalHazard.cs

```diff
@@ -0,0 +1,85 @@
+    /// <br>Adds the better effect logic.</br>
+    /// </summary>
+    /// <seealso cref="StayingOnSinkholeEnvironmentalHazard"/>
+    /// <seealso cref="SinkholeEffectFix"/>
+    [HarmonyPatch(typeof(EnvironmentalHazard), nameof(EnvironmentalHazard.OnExit))]
+    internal static class ExitingSinkholeEnvironmentalHazard
+    {
+        private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions, ILGenerator generator)
+        {
+            List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Shared.Rent(instructions);
+
+            Label cnt = generator.DefineLabel();
+
+            // We add type check because SinkholeEnvironmentalHazard dont override OnExit method
+            // Without type check, the TantrumEnvironmentalHazard::OnExit event will be called several times
+            newInstructions.InsertRange(newInstructions.Count - 1, new[]
+            {
+                new(OpCodes.Ldarg_0),
+                new(OpCodes.Isinst, typeof(SinkholeEnvironmentalHazard)),
+                new(OpCodes.Brfalse_S, cnt),
+                new(OpCodes.Ldarg_1),
+                new(OpCodes.Call, Method(typeof(API.Features.Player), nameof(API.Features.Player.Get), new[] { typeof(ReferenceHub) })),
+                new(OpCodes.Ldarg_0),
+                new(OpCodes.Ldc_I4_1),
+                new(OpCodes.Newobj, GetDeclaredConstructors(typeof(ExitingEnvironmentalHazardEventArgs))[0]),
+                new(OpCodes.Dup),
+                new(OpCodes.Call, Method(typeof(Handlers.Player), nameof(Handlers.Player.OnExitingEnvironmentalHazard))),
+                new(OpCodes.Callvirt, PropertyGetter(typeof(ExitingEnvironmentalHazardEventArgs), nameof(ExitingEnvironmentalHazardEventArgs.IsAllowed))),
+
+                // If IsAllowed == false, we dont remove RefHub from AffectedPlayers, only dont removing effect
+                new(OpCodes.Brfalse_S, cnt),
+
+                // null check
+                new(OpCodes.Ldarg_1),
+                new(OpCodes.Brfalse_S, cnt),
+
+                // SCP check
+                new(OpCodes.Ldarg_1),
+                new(OpCodes.Ldfld, Field(typeof(ReferenceHub), nameof(ReferenceHub.characterClassManager))),
+                new(OpCodes.Callvirt, Method(typeof(CharacterClassManager), nameof(CharacterClassManager.IsAnyScp))),
+                new(OpCodes.Brtrue_S, cnt),
+
+                // exit effect for 1 second, cause we disable effect in OnStay
+                new(OpCodes.Ldarg_1),
+                new(OpCodes.Ldfld, Field(typeof(ReferenceHub), nameof(ReferenceHub.playerEffectsController))),
+                new(OpCodes.Ldc_R4, 1f),
+                new(OpCodes.Ldc_I4_0),
+                new(OpCodes.Callvirt, Method(typeof(PlayerEffectsController), nameof(PlayerEffectsController.EnableEffect), new[] { typeof(float), typeof(bool) }, new[] { typeof(SinkHole) })),
+
+                new CodeInstruction(OpCodes.Nop).WithLabels(cnt),
```

Comment:
How about? (I assume the last instruction is a ret)

newInstructions[newInstructions.Count - 1].withLabels(cnt)

Reply:
Yeah I  agree, `Nop` isn't needed.

Reply:
There was a little different logic here, and Nop was needed, but yes, it is not needed now, I will remove it tomorrow